# Integrantes:
Angela Patricia Ordoñez Castro    
Cod:201628709

Juan Sebastián Garés Carranza     
Cod:201628949

# Exercise 03

## Data preparation and model evaluation exercise with credit scoring

Banks play a crucial role in market economies. They decide who can get finance and on what terms and can make or break investment decisions. For markets and society to function, individuals and companies need access to credit. 

Credit scoring algorithms, which make a guess at the probability of default, are the method banks use to determine whether or not a loan should be granted. This competition requires participants to improve on the state of the art in credit scoring, by predicting the probability that somebody will experience financial distress in the next two years. [Dataset](https://www.kaggle.com/c/GiveMeSomeCredit)

Attribute Information:

|Variable Name	|	Description	|	Type|
|----|----|----|
|SeriousDlqin2yrs	|	Person experienced 90 days past due delinquency or worse 	|	Y/N|
|RevolvingUtilizationOfUnsecuredLines	|	Total balance on credit divided by the sum of credit limits	|	percentage|
|age	|	Age of borrower in years	|	integer|
|NumberOfTime30-59DaysPastDueNotWorse	|	Number of times borrower has been 30-59 days past due |	integer|
|DebtRatio	|	Monthly debt payments	|	percentage|
|MonthlyIncome	|	Monthly income	|	real|
|NumberOfOpenCreditLinesAndLoans	|	Number of Open loans |	integer|
|NumberOfTimes90DaysLate	|	Number of times borrower has been 90 days or more past due.	|	integer|
|NumberRealEstateLoansOrLines	|	Number of mortgage and real estate loans	|	integer|
|NumberOfTime60-89DaysPastDueNotWorse	|	Number of times borrower has been 60-89 days past due |integer|
|NumberOfDependents	|	Number of dependents in family	|	integer|


Read the data into Pandas

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 500)
import zipfile
with zipfile.ZipFile('../datasets/KaggleCredit2.csv.zip', 'r') as z:
    f = z.open('KaggleCredit2.csv')
    data = pd.io.parsers.read_table(f, sep=',')

data.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines   age  \
0           0                 1                              0.766127  45.0   
1           1                 0                              0.957151  40.0   
2           2                 0                              0.658180  38.0   
3           3                 0                              0.233810  30.0   
4           4                 0                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

# Exercise 3.1

Input the missing values of the Age and Number of Dependents 

# Solución:
identificamos cuantos valores perdidos tienen las variables Age y Number of Dependents:
data.isnull().sum()


In [2]:
data.isnull().sum()

Unnamed: 0                                 0
SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                     4267
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                              0
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                      4267
dtype: int64

Se identifican 4267 valores perdidos en la variable age y 4267 en la variable NumberOfDependents

In [3]:
data.age.fillna(data.age.median(), inplace=True)
data.NumberOfDependents.fillna(data.NumberOfDependents.median(), inplace=True)
data.isnull().sum()

Unnamed: 0                              0
SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

Mediante la mediana de la edad y número de dependientes, se imputaron los valores perdidos de estas variable, la base de datos no contiene más valores perdidos

In [4]:
y = data['SeriousDlqin2yrs']
X = data.drop(['Unnamed: 0','SeriousDlqin2yrs'], axis=1)

In [5]:
X.head()

RevolvingUtilizationOfUnsecuredLines   age  \
0                              0.766127  45.0   
1                              0.957151  40.0   
2                              0.658180  38.0   
3                              0.233810  30.0   
4                              0.907239  49.0   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                   2.0   0.802982         9120.0   
1                                   0.0   0.121876         2600.0   
2                                   1.0   0.085113         3042.0   
3                                   0.0   0.036050         3300.0   
4                                   1.0   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                             13.0                      0.0   
1                              4.0                      0.0   
2                              2.0                      1.0   
3                              5.0                      0.0   
4                              7.0                      0.0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                           6.0                                   0.0   
1                           0.0                                   0.0   
2                           0.0                                   0.0   
3                           0.0                                   0.0   
4                           1.0                                   0.0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

# Exercise 3.2

From the set of features

Select the features that maximize the **F1Score** the model using K-Fold cross-validation

In [6]:
X.shape

(112915, 10)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.model_selection import KFold
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

# list of values to try for max_features
feature_range = range(1, len(data.columns)+1)

# list to store the average RMSE for each value of max_features
F1_scores = []

# use 10-fold cross-validation with each value of max_features (WARNING: SLOW!)
for feature in feature_range:
    rfreg = RandomForestClassifier(n_estimators=150, max_features=feature, random_state=1, n_jobs=-1)
    F1_scores = cross_val_score(rfreg, X, y, cv=5, scoring='f1')
    
#X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

#rfreg = RandomForestClassifier(n_estimators=150, max_features=10, max_depth=5, oob_score=True, random_state=1)
#rfreg.fit(X_train, y_train)

#from sklearn.feature_selection import SelectFromModel
#print(SelectFromModel(rfreg, threshold=0.1, prefit=True).transform(X_train).shape)
#print(SelectFromModel(rfreg, threshold='mean', prefit=True).transform(X_train).shape)
#print(SelectFromModel(rfreg, threshold='median', prefit=True).transform(X_train).shape)
    

In [ ]:
plt.plot(feature_range, RMSE_scores)
plt.xlabel('max_features')
plt.ylabel('RMSE (lower is better)')

In [ ]:
# show the best RMSE and the corresponding max_features
sorted(zip(RMSE_scores, feature_range))[0]

In [ ]:
# max_features=8 is best and n_estimators=150 is sufficiently large
rfreg = RandomForestClassifier(n_estimators=150, max_features=8, max_depth=3, oob_score=True, random_state=1)
rfreg.fit(X, y)

In [ ]:
# define features: exclude career statistics (which start with "C") and the response (Salary)
feature_cols = data.columns[data.columns.str.startswith('C') == False].drop(['Unnamed: 0','SeriousDlqin2yrs'])
feature_cols

In [ ]:
# compute feature importances
pd.DataFrame({'feature':feature_cols, 'importance':rfreg.feature_importances_}).sort_values('importance')

Las variables que maximizan el f1_score son RevolvingUtilizationOfUnsecuredLines, NumberOfTime30-59DaysPastDueNotWors y MonthlyIncome

# Exercise 3.3

Now which is the best set of features selected by AUC

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.model_selection import KFold
from sklearn.feature_selection import SelectFromModel
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

# list of values to try for max_features
feature_range = range(1, len(data.columns)+1)

# list to store the average RMSE for each value of max_features
F1_scores = []

# use 10-fold cross-validation with each value of max_features (WARNING: SLOW!)
for feature in feature_range:
    rfreg = RandomForestClassifier(n_estimators=150, max_features=feature, random_state=1, n_jobs=-1)
    F1_scores = cross_val_score(rfreg, X, y, cv=5, scoring='roc_auc')
    
#X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

#rfreg = RandomForestClassifier(n_estimators=150, max_features=10, max_depth=5, oob_score=True, random_state=1)
#rfreg.fit(X_train, y_train)

#from sklearn.feature_selection import SelectFromModel
#print(SelectFromModel(rfreg, threshold=0.1, prefit=True).transform(X_train).shape)
#print(SelectFromModel(rfreg, threshold='mean', prefit=True).transform(X_train).shape)
#print(SelectFromModel(rfreg, threshold='median', prefit=True).transform(X_train).shape)
    

In [ ]:
plt.plot(feature_range, RMSE_scores)
plt.xlabel('max_features')
plt.ylabel('RMSE (lower is better)')

In [ ]:
# show the best RMSE and the corresponding max_features
sorted(zip(RMSE_scores, feature_range))[0]

In [ ]:
# max_features=8 is best and n_estimators=150 is sufficiently large
rfreg = RandomForestClassifier(n_estimators=150, max_features=8, max_depth=3, oob_score=True, random_state=1)
rfreg.fit(X, y)

In [ ]:
# define features: exclude career statistics (which start with "C") and the response (Salary)
feature_cols = data.columns[data.columns.str.startswith('C') == False].drop(['Unnamed: 0','SeriousDlqin2yrs'])
feature_cols

In [ ]:
# compute feature importances
pd.DataFrame({'feature':feature_cols, 'importance':rfreg.feature_importances_}).sort_values('importance')